edisgo/network/topology.py

```diff
@@ -2399,61 +2455,199 @@ def connect_to_lv_based_on_geolocation(
             :attr:`~.network.topology.Topology.loads_df` or
             :attr:`~.network.topology.Topology.storage_units_df`, depending on component
             type.
-
         """
 
+        def validate_voltage_level(voltage_level):
+            if voltage_level not in [6, 7]:
+                raise ValueError(
+                    f"Voltage level must either be 6 or 7 but given voltage level "
+                    f"is {voltage_level}."
+                )
+
+        def get_add_function(comp_type):
+            add_func_map = {
+                "generator": self.add_generator,
+                "charging_point": self.add_load,
+                "heat_pump": self.add_load,
+                "storage_unit": self.add_storage_unit,
+            }
+            add_func = add_func_map.get(comp_type)
+            if add_func is None:
+                raise ValueError(
+                    f"Provided component type {comp_type} is not valid. Must either be "
+                    f"'generator', 'charging_point', 'heat_pump' or 'storage_unit'."
+                )
+            return add_func
+
+        def handle_voltage_level_6():
+            # get substations and MV buses
+            substations = self.buses_df.loc[self.transformers_df.bus1.unique()]
+            if allow_mv_connection:
+                mv_buses = self.buses_df.loc[self.mv_grid.buses_df.index]
+            else:
+                mv_buses = pd.DataFrame()
+            all_buses = pd.concat([substations, mv_buses])
+            # calculate distance to possible buses
+            target_bus, target_bus_distance = geo.find_nearest_bus(
+                geolocation, all_buses
+            )
+            if target_bus in substations.index:
+                # if distance is larger than allowed, create new bus and connect to
+                # station via a new line
+                if target_bus_distance > max_distance_from_target_bus:
+                    bus = self._connect_to_lv_bus(
+                        edisgo_object, target_bus, comp_type, comp_data
+                    )
+                else:
+                    mvlv_subst_id = self.buses_df.loc[target_bus].loc["lv_grid_id"]
+                    comp_data["mvlv_subst_id"] = mvlv_subst_id
+                    comp_name = self.connect_to_lv(edisgo_object, comp_data, comp_type)
```

Comment:
I think here it makes more sense to directly use the add_func, since the bus to connect to is already known.